File: app/src/main/java/net/tlalka/android/fiszki/injections/modules/SessionModule.java
package net.tlalka.android.fiszki.injections.modules;

import android.content.Context;
import dagger.Module;
import dagger.Provides;
import net.tlalka.android.fiszki.injections.annotations.SessionScope;
import net.tlalka.android.fiszki.models.db.DbHelper;
import net.tlalka.android.fiszki.models.db.DbManager;
import net.tlalka.android.fiszki.services.AssetsHelper;
import net.tlalka.android.fiszki.services.StorageHelper;

@Module
public class SessionModule {

    private final Context context;

    public SessionModule(Context context) {
        this.context = context;
    }

    @Provides
    @SessionScope
    Context context() {
        return this.context;
    }

    @Provides
    @SessionScope
    DbHelper dbHelper() {
        return DbManager.getHelper(this.context);
    }

    @Provides
    @SessionScope
    StorageHelper storageHelper() {
        return new StorageHelper(this.context);
    }

    @Provides
    @SessionScope
    AssetsHelper assetsHelper() {
        return new AssetsHelper(this.context);
    }
}

File: app/src/main/kotlin/net/tlalka/fiszki/core/modules/SessionModule.kt
package net.tlalka.fiszki.core.modules

import android.content.Context

import dagger.Module
import dagger.Provides
import net.tlalka.fiszki.core.annotations.SessionScope
import net.tlalka.fiszki.model.dao.ClusterDao
import net.tlalka.fiszki.model.dao.LessonDao
import net.tlalka.fiszki.model.dao.WordDao
import net.tlalka.fiszki.model.db.DbHelper
import net.tlalka.fiszki.model.db.DbManager
import net.tlalka.fiszki.model.helpers.AssetsHelper
import net.tlalka.fiszki.model.helpers.StorageHelper

@Module
class SessionModule(@get:Provides @get:SessionScope val context: Context) {

    @Provides
    @SessionScope
    fun getDbHelper(): DbHelper {
        return DbManager.getHelper(context)
    }

    @Provides
    @SessionScope
    fun getLessonDao(dbHelper: DbHelper): LessonDao {
        return dbHelper.getLessonDao()
    }

    @Provides
    @SessionScope
    fun getClusterDao(dbHelper: DbHelper): ClusterDao {
        return dbHelper.getClusterDao()
    }

    @Provides
    @SessionScope
    fun getWordDao(dbHelper: DbHelper): WordDao {
        return dbHelper.getWordDao()
    }

    @Provides
    @SessionScope
    fun getStorageHelper(): StorageHelper {
        return StorageHelper(context)
    }

    @Provides
    @SessionScope
    fun getAssetsHelper(): AssetsHelper {
        return AssetsHelper(context)
    }
}
